core/shared/src/main/scala/fs2/Chunk.scala

```diff
@@ -945,19 +957,37 @@ object Chunk
       else new Queue(chunks :+ that, size + that.size)
 
     /** Prepends a chunk to the start of this chunk queue. */
-    def +:[O2 >: O](c: Chunk[O2]): Queue[O2] = new Queue(c +: chunks, c.size + size)
+    def +:[O2 >: O](c: Chunk[O2]): Queue[O2] =
+      if (c.isEmpty) this else new Queue(c +: chunks, c.size + size)
 
     /** Appends a chunk to the end of this chunk queue. */
-    def :+[O2 >: O](c: Chunk[O2]): Queue[O2] = new Queue(chunks :+ c, size + c.size)
+    def :+[O2 >: O](c: Chunk[O2]): Queue[O2] =
+      if (c.isEmpty) this else new Queue(chunks :+ c, size + c.size)
 
     def apply(i: Int): O = {
       if (i < 0 || i >= size) throw new IndexOutOfBoundsException()
-      def go(chunks: SQueue[Chunk[O]], offset: Int): O = {
-        val head = chunks.head
-        if (offset < head.size) head(offset)
-        else go(chunks.tail, offset - head.size)
+      if (i == 0) chunks.head(0)
+      else if (i == size - 1) chunks.last.last.get
+      else {
+        val table = accumulatedLengths
+        var lo = 0
+        var hi = table.length - 1
+        var mid = (hi - lo) / 2
+        var chunk: Chunk[O] = null
+        while (lo <= hi) {
+          val delta = hi - lo
+          mid = lo + (delta / 2)
+          val entry = table(mid)
+          val accLen = entry._1
+          chunk = entry._2
+          if (i >= accLen) lo = if (delta <= 1) mid + 1 else mid
+          else if (i >= accLen - chunk.size) lo = hi + 1
+          else hi = mid
+        }
```

Comment:
In line with the above, once the array of accumulated lengths is separate, perhaps we can use a Java utility for binary search. https://docs.oracle.com/javase/7/docs/api/java/util/Arrays.html#binarySearch(int[],%20int,%20int,%20int)

Reply:
Note we're not searching for a specific key so I don't think we can reuse `binarySearch`.

Reply:
Right... the point of the binary search is to give you the last position, `j`, in the array of accumulated lengths, which is smaller than the intended element position `i`. That would be where the chunk containing element `i` resides. 

So, for chunks of lengths `5, 3, 6, 12, 4`, your accumulated lenghts would be `5, 8, 15, 26, 30`, so looking for element 7 should yield you to the second chunk, and looking for element 29 should take you to the last one. The Java utility would remove the loop, but the result would have to be adjusted...

Reply:
Understood but I still don't see how that's possible to implement with `Arrays.binarySearch`. Can you give it a try?